apis/python/src/tiledbsoma/io/shaping.py

```diff
@@ -0,0 +1,546 @@
+        uri: The URI of a SOMA :class:`Experiment``.
+        nobs: The desired new shape of the experiment's ``obs`` dataframe.
+        nvars: The desired new shapes of the experiment's ``var`` dataframes.
+            This should be a dict from measurement name to shape, e.g.
+            ``{"RNA": 10000, "raw": 20000}``.
+        verbose: If ``True``, produce per-array output as the upgrade runs.
+        check_only: If ``True``,  don't apply the upgrades, but show what would
+            be attempted, and show why each one would fail.
+        context: Optional :class:`SOMATileDBContext``.
+    """
+    args: SizingArgs = dict(
+        nobs=nobs,
+        nvars=nvars,
+        ms_name=None,
+        coll_name=None,
+        verbose=verbose,
+        check_only=check_only,
+        context=context,
+        output_handle=output_handle,
+    )
+
+    # Do an early check on the nvars keys vs the experiment's
+    # measurent names. This isn't a can-do status for the experiment;
+    # it's a failure of the user's arguments.
+    with tiledbsoma.Experiment.open(uri) as exp:
+        arg_keys = sorted(list(set(nvars.keys())))
+        ms_keys = sorted(list(set(exp.ms.keys())))
+        if arg_keys != ms_keys:
+            raise ValueError(
+                f"resize_experiment: provided nvar keys {arg_keys} do not match experiment keys {ms_keys}"
+            )
+
+    ok = _treewalk(
+        uri,
+        visitor=_leaf_visitor_resize,
+        args=args,
+    )
+    return ok
+
+
+def _treewalk(
+    uri: str,
+    *,
+    node_name: Optional[str] = None,
+    visitor: Any,
+    args: SizingArgs,
+) -> bool:
+    retval = True
+    with tiledbsoma.open(uri) as item:
+
+        if isinstance(item, tiledbsoma.Experiment):
+            if "obs" in item:
+                ok = _treewalk(
+                    item["obs"].uri, node_name="obs", visitor=visitor, args=args
+                )
+                retval = retval and ok
+            if "ms" in item:
+                ok = _treewalk(
+                    item["ms"].uri, node_name="ms", visitor=visitor, args=args
+                )
+                retval = retval and ok
+
+        elif isinstance(item, tiledbsoma.Measurement):
+            if "var" in item:
+                ok = _treewalk(
+                    item["var"].uri, node_name="var", visitor=visitor, args=args
+                )
+                retval = retval and ok
+
+            for coll_name in ["X", "obsm", "obsp", "varm", "varp"]:
+                if coll_name in item:
+
+                    args = dict(
+                        nobs=args["nobs"],
+                        nvars=args["nvars"],
+                        ms_name=args["ms_name"],
+                        coll_name=coll_name,
+                        verbose=args["verbose"],
+                        check_only=args["check_only"],
+                        context=args["context"],
+                        output_handle=args["output_handle"],
+                    )
```

Comment:
(I think you can do this)

```suggestion
                    args = dict(coll_name=coll_name, **args)
```